Issues/api tweaks

aerych opened the pull request:
This is a housekeeping patch addressing a few annoyances with our XMLRPC API calls about which I have an axe to grind. 

This patch:
- Reduces the number of WordPressXMLRPCApi instances created from sign up to viewing the post list from 10 to 3, and only a single instance is created during subsequent launches. The two extra instances during signup are used for blog detection and validation.
- Removes the staggered sync request methods we were using to avoid issues with basic http auth with some hosts. We should no longer need this work around if we're reusing a single api instance. The max concurrent request setting should be enough, but we needs through testing. @maxme, @jleandroperez can you help me with some test servers again to verify this works? 
- Separates syncing blog meta data (options, post formats, and categories) into its own method that runs in the background. 
- Updates the editor to sync post meta only (and not all post content) when being launched from the tab bar. This reduces the network calls made from 7 to 3 avoiding some waste, potentially addresses an edge case where local edits to a post/page were being erased(!), and helps reduce the risk of running afoul of some host's overly aggressive request throttling. (cc @bummytime) 
- Removes the callback blocks from the blog sync method since they were never used to begin with. 
- Adds some logging for failed requests. 

Needs review: @koke

koke commented on WordPress/Classes/Services/BlogService.h:
```diff
@@ -58,15 +58,19 @@
+ post formats, blog options, and categories. Also checks if the blog is multi-author.
+ Used for instances where the entire blog should be refreshed or initially downloaded.
+ */
+- (void)backgroundSyncBlog:(Blog *)blog;
```
What if we just stopped doing this? I think we want to keep options synced, and we want formats/categories to be there when editing a post. Everything else we could just load on demand.

Also, why the name change? I'm fine with it, but just curious.

aerych replied:
> What if we just stopped doing this?

I'm a fan of this idea, honestly. A short load time the first time the app is launched shouldn't be a big deal and even if the user happens to have a lot of latency they probably would still be waiting on the initial sync to finish by the time the signed in and tapped to view their post list.

> Also, why the name change? I'm fine with it, but just curious.

I might have been influenced by my perception of how the method is used: fire-and-forget, callbacks never used. If it doesn't make sense I can switch it back.

aerych commented:
Kudos to @jleandroperez for setting up a test server. :)

jleandroperez commented:
Anytime Eric! :smile:

jleandroperez commented:
Just stepping by to confirm that this PR works like a charm with Basic Auth. Awesome work there @aerych!

aerych commented:
@koke, I've removed the content syncing.  I also went a head and changed the method name. After some thought it didn't make much sense for it to imply it was a "background" operation.
Ready for round two!

koke commented on WordPress/Classes/Services/BlogSyncFacade.m:
```diff
@@ -49,7 +49,7 @@ - (void)syncBlogWithUsername:(NSString *)username
     blog.options = options;
     [[ContextManager sharedInstance] saveContext:context];
-    [blogService syncBlog:blog success:nil failure:nil];
+    [blogService syncBlog:blog];
```
Since we already have the `options` I think we could skip the sync here as well.

aerych replied:
Makes sense. We'd still want to do the multi author check but another sync is triggered when the user views the blog detail so not syncing here should be fine.

koke commented on WordPress/Classes/Services/BlogService.m:
```diff
@@ -230,144 +230,52 @@ - (void)syncPostFormatsForBlog:(Blog *)blog
-- (void)syncBlog:(Blog *)blog
-         success:(void (^)())success
-         failure:(void (^)(NSError *error))failure
+- (void)backgroundSyncMetaForBlog:(Blog *)blog
```
It looks like nothing is calling this

aerych replied:
:facepalm: Nothing is, and it should have been removed.

koke commented:
Other than the two comments, code looks good. Is there anything specific I should be testing?

aerych commented:
> Is there anything specific I should be testing

My biggest concern was how the changes would play with basic http auth, but I think its going to be okay after testing with Jorge.  Beyond that there wasn't anything in particular, just some general testing to make sure nothing was breaking with the patch.

koke commented:
I definitely think we should kill the `syncBlog` after login. Maybe even the one when the blog details view is displayed: if I see that screen because I'm going to moderate comments, I don't need to sync post formats.

I tested this on a *very slow* connection, so I might be biased by the long wait times. I logged in to my self hosted site and immediately navigated to the posts list.

With the current implementation, it took 2.5 minutes until I saw a post:

![screen shot 2015-07-13 at 16 58 04](https://cloud.githubusercontent.com/assets/8739/8652972/3f6f5dc0-2982-11e5-92b2-a32912f24351.png)

Removing the `syncBlog` on `BlogSyncFacade` and `BlogDetailsViewController`, it takes _only_ 1.5 minutes:

![screen shot 2015-07-13 at 17 00 15](https://cloud.githubusercontent.com/assets/8739/8653048/902b4cba-2982-11e5-86be-6a32292f42d8.png)

These are very bad network conditions, and it might be even better than the previous scenario. But I was testing this because I was worried that all that background syncing might get in the way of getting the content the user wants, and I think that still applies here. If you notice on the first flow, the `wp.getPosts` call isn't even posted to the server until I had been looking at the screen for about 30 seconds. Since we don't have any checks to prevent duplicate requests going out, I can imagine an impatient user navigating in and out of views seeing if *anything* loads at all, adding network requests that just make everything worse.

Now I know you're going to yell "Cancel!", to which I'll reply "Prioritize traffic!" :stuck_out_tongue: 

Anyway, I think this might require some extra thought before we serialize requests more (I know we're not exactly serializing, but setting a maximum concurrency of 3 and sending 4 requests on every syncBlog means it's blocking the network for everything else)

aerych commented:
> if I see that screen because I'm going to moderate comments, I don't need to sync post formats.

True. Maybe we should move categories and post format syncing else where, or at least time-gate them so they aren't triggered repeatedly within a certain time span.  

> Now I know you're going to yell "Cancel!",

The tune I'm singing today is we probably don't need to cancel. It might be enough to be able to ignore ;D

aerych commented:
@koke round three?

koke commented:
:shipit:

aerych commented:
Thanks Koke!
